src/ssl.c

```diff
@@ -73,41 +81,165 @@ ssl_startup(void)
+	if (strcasecmp(hostname, common_name_str) == 0) {
+		return MatchFound;
+	} else {
+		return MatchNotFound;
+	}
+}
+
+static validate_result
+matches_subject_alternative_name(const char *hostname, const X509 *server_cert)
+{
+	validate_result result = MatchNotFound;
+	int i;
+	int san_names_nb = -1;
+	STACK_OF(GENERAL_NAME) *san_names = NULL;
+
+	// Try to extract the names within the SAN extension from the certificate
+	san_names = X509_get_ext_d2i((X509 *) server_cert, NID_subject_alt_name, NULL, NULL);
+	if (san_names == NULL) {
+		return NoSANPresent;
+	}
+	san_names_nb = sk_GENERAL_NAME_num(san_names);
+
+	// Check each name within the extension
+	for (i = 0; i < san_names_nb; i++) {
+		const GENERAL_NAME *current_name = sk_GENERAL_NAME_value(san_names, i);
+
+		if (current_name->type == GEN_DNS) {
+			// Current name is a DNS name, let's check it
+#if OPENSSL_VERSION_NUMBER < 0x10101000L
+			char *dns_name = (char *) ASN1_STRING_data(current_name->d.dNSName);
+#else
+			char *dns_name = (char *) ASN1_STRING_get0_data(current_name->d.dNSName);
+#endif
+			// Make sure there isn't an embedded NUL character in the DNS name
+			if ((size_t) ASN1_STRING_length(current_name->d.dNSName) != strlen(dns_name)) {
+				result = MalformedCertificate;
+				break;
+			} else {
+				// Compare expected hostname with the DNS name
+				if (strcasecmp(hostname, dns_name) == 0) {
+					result = MatchFound;
+					break;
+				}
+			}
+		}
+	}
+	sk_GENERAL_NAME_pop_free(san_names, GENERAL_NAME_free);
+
+	return result;
+}
+
+
+validate_result
+validate_hostname(const char *hostname, const X509 *server_cert)
+{
+	validate_result result;
+
+	if ((hostname == NULL) || (server_cert == NULL)) {
+		return Error;
+	}
+
+	// First try the Subject Alternative Names extension
+	result = matches_subject_alternative_name(hostname, server_cert);
+	if (result == NoSANPresent) {
+		// Extension was not found: try the Common Name
+		result = matches_common_name(hostname, server_cert);
+	}
+
+	return result;
+}
+
 SSL *
 ssl_connect(int fd, char *hostname)
 {
-
+	X509 *server_cert;
 	SSL_CTX *ssl_ctx;
 	SSL *ssl;
 
 	ssl_startup();
 
 	ssl_ctx = SSL_CTX_new(SSLv23_client_method());
-	ssl = SSL_new(ssl_ctx);
 	if (!conf->insecure) {
-		SSL_set_hostflags(ssl, X509_CHECK_FLAG_NO_PARTIAL_WILDCARDS);
-		SSL_set1_host(ssl, hostname);
 		SSL_CTX_set_default_verify_paths(ssl_ctx);
 		SSL_CTX_set_verify(ssl_ctx, SSL_VERIFY_PEER, NULL);
 	}
 	SSL_CTX_set_mode(ssl_ctx, SSL_MODE_AUTO_RETRY);
 
+	ssl = SSL_new(ssl_ctx);
 	SSL_set_fd(ssl, fd);
 	SSL_set_tlsext_host_name(ssl, hostname);
 
 	int err = SSL_connect(ssl);
 	if (err <= 0) {
 		fprintf(stderr, _("SSL error: %s\n"),
 			ERR_reason_error_string(ERR_get_error()));
+		SSL_CTX_free(ssl_ctx);
 		return NULL;
 	}
 
 	err = SSL_get_verify_result(ssl);
 	if (err != X509_V_OK) {
 		fprintf(stderr, _("SSL error: Certificate error"));
+		SSL_CTX_free(ssl_ctx);
+		return NULL;
+	}
+
+	server_cert =  SSL_get_peer_certificate(ssl);
```

Comment:
this could be acquired inside `validate_hostname`.

Reply:
I feel it makes sense here to differentiate between the two error cases: certificate not found and hostname verification failure. But I will defer to your direction, let me know what you'd like me to do.